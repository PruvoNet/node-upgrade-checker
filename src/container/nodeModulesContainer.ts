import * as typeorm from 'typeorm';
import * as fs from 'fs';
import * as pacote from 'pacote';
import * as simpleGit from 'simple-git/promise';
import * as spawn from 'cross-spawn';
import { interfaces } from 'inversify';
import Bind = interfaces.Bind;

export type Spawn = typeof spawn;
export type Pacote = typeof pacote;
export type FS = typeof fs;
export type TypeOrm = typeof typeorm;
export type SimpleGitFn = typeof simpleGit;
export const TYPES = {
  TypeOrm: Symbol.for(`TypeOrm`),
  FS: Symbol.for(`FS`),
  Pacote: Symbol.for(`Pacote`),
  Spawn: Symbol.for(`Spawn`),
  SimpleGit: Symbol.for(`SimpleGit`),
};

export const nodeModulesBinder = (bind: Bind): void => {
  bind<TypeOrm>(TYPES.TypeOrm).toConstantValue(typeorm);
  bind<FS>(TYPES.FS).toConstantValue(fs);
  bind<Pacote>(TYPES.Pacote).toConstantValue(pacote);
  bind<Spawn>(TYPES.Spawn).toConstantValue(spawn);
  bind<SimpleGitFn>(TYPES.SimpleGit).toConstantValue(simpleGit);
};
