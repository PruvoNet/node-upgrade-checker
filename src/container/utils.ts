import { METADATA_KEY, namedConstraint, taggedConstraint, interfaces } from 'inversify';
import Request = interfaces.Request;
import Abstract = interfaces.Abstract;

export type Constraint = (request: Request) => boolean;

export const namedOrMultiConstraint = (symbol: symbol, target: Abstract<unknown>): Constraint => {
  const tagConstraint = namedConstraint(symbol);
  const multiInjectConstraint = taggedConstraint(METADATA_KEY.MULTI_INJECT_TAG)(target);
  return (request: Request): boolean => {
    return multiInjectConstraint(request) || tagConstraint(request);
  };
};
