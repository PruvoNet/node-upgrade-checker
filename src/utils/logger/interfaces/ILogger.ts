export enum LogLevel {
  ERROR,
  WARN,
  LOG,
  INFO,
  DEBUG,
  TRACE,
  SILENT,
}

export interface ILogger {
  fatal(...args: unknown[]): void;

  error(...args: unknown[]): void;

  warn(...args: unknown[]): void;

  log(...args: unknown[]): void;

  info(...args: unknown[]): void;

  success(...args: unknown[]): void;

  debug(...args: unknown[]): void;

  trace(...args: unknown[]): void;

  isDebugEnabled(): boolean;
  isTraceEnabled(): boolean;
  isInfoEnabled(): boolean;
  isWarnEnabled(): boolean;
  isLevelEnabled(level: LogLevel): boolean;
}
