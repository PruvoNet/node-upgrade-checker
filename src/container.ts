import {Container} from 'inversify';
import {gitContainerModule} from './utils/git';
import {runnerContainerModule} from './utils/runner';
import {npmContainerModule} from './utils/npm';
import {ciResolverContainerModule} from './resolvers/ciResolver';
import {dbContainerModule} from './db';
import {cacheResolverContainerModule} from './resolvers/cacheResolver';
import {testResolverContainerModule} from './resolvers/testResolver';

export const container = new Container();

container.load(gitContainerModule);
container.load(runnerContainerModule);
container.load(npmContainerModule);
container.load(ciResolverContainerModule);
container.load(dbContainerModule);
container.load(cacheResolverContainerModule);
container.load(testResolverContainerModule);
