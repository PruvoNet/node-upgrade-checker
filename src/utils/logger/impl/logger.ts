import { Consola, ConsolaOptions } from 'consola';
import { ILogger, LogLevel } from '../interfaces/ILogger';

export class Logger extends Consola implements ILogger {
  constructor(options: ConsolaOptions) {
    super(options);
  }

  public isLevelEnabled(level: LogLevel): boolean {
    return this.level >= level;
  }

  public isTraceEnabled(): boolean {
    return this.isLevelEnabled(LogLevel.TRACE);
  }

  public isDebugEnabled(): boolean {
    return this.isLevelEnabled(LogLevel.DEBUG);
  }

  public isInfoEnabled(): boolean {
    return this.isLevelEnabled(LogLevel.INFO);
  }

  public isWarnEnabled(): boolean {
    return this.isLevelEnabled(LogLevel.WARN);
  }
}
