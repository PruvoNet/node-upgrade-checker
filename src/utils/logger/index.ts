import { interfaces } from 'inversify';
import Bind = interfaces.Bind;
import { LoggerFactory } from './impl/loggerFactory';
import { ILoggerFactory } from './interfaces/loggerFactory';
import { LoggerSettings } from './impl/loggerSettings';
import { ILoggerSettings } from './interfaces/loggerSettings';

export const loggerModuleBinder = (bind: Bind): void => {
  bind<ILoggerSettings>(ILoggerSettings)
    .to(LoggerSettings)
    .inSingletonScope();
  bind<ILoggerFactory>(ILoggerFactory)
    .to(LoggerFactory)
    .inSingletonScope();
};

export * from './interfaces/loggerFactory';
export * from './interfaces/loggerSettings';
export * from './impl/loggerSettings';
