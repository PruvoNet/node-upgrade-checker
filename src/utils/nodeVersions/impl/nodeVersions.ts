import { inject, injectable } from 'inversify';
import { INodeVersions, INodeVersionsLtsOptions, INodeVersionsLatestOptions } from '../interfaces/INodeVersions';
import { Moment } from 'moment';
import moment = require('moment');
import { memoize } from '../../memoize/memoize';
import { Axios, TYPES } from '../../../container/nodeModulesContainer';
import { ILoggerFactory } from '../../logger';
import { ILogger } from '../../logger/interfaces/ILogger';
import { O } from 'ts-toolbelt';

const NODE_RELEASE_FILE = `https://raw.githubusercontent.com/nodejs/Release/master/schedule.json`;
const dateFormat = `YYYY-MM-DD`;

interface IRemoteNodeVersion {
  start: string;
  end: string;
  codename?: string;
  lts?: string;
}

type RemoteNodeVersions = Record<string, IRemoteNodeVersion>;

interface INodeVersion {
  from: Moment;
  ltsFrom: Moment;
  to: Moment;
  version: string;
  codename: string;
}

const dateSorter = (key: O.SelectKeys<INodeVersion, Moment>) => (a: INodeVersion, b: INodeVersion): number => {
  return a[key].valueOf() - b[key].valueOf();
};

@injectable()
export class NodeVersions extends INodeVersions {
  private readonly logger: ILogger;
  constructor(@inject(TYPES.Axios) private readonly axios: Axios, loggerFactory: ILoggerFactory) {
    super();
    this.logger = loggerFactory.getLogger(`Node Versions`);
  }

  @memoize(({ codename }: INodeVersionsLtsOptions): string => codename.toLowerCase())
  public async resolveLtsVersion({ codename }: INodeVersionsLtsOptions): Promise<string | undefined> {
    codename = codename.toLowerCase();
    this.logger.debug(`Resolving LTS version of ${codename}`);
    const versions = await this.getAllVersions();
    const relevantVersions = versions
      .filter((nodeVersion: INodeVersion) => nodeVersion.codename === codename)
      .map((nodeVersion: INodeVersion) => nodeVersion.version);
    const lts = relevantVersions[0];
    if (!lts) {
      this.logger.error(`Failed to resolve LTS version of ${codename}`);
      return;
    }
    this.logger.debug(`Resolved LTS version of ${codename} to ${lts}`);
    return lts;
  }

  @memoize(({ date }: INodeVersionsLatestOptions): string => date.toJSON())
  public async resolveLatestLtsVersion({ date }: INodeVersionsLatestOptions): Promise<string | undefined> {
    this.logger.debug(`Resolving latest lts version in ${date.toJSON()}`);
    const versions = await this.getAllVersions();
    const relevantVersions = versions
      .filter(
        (nodeVersion: INodeVersion) => nodeVersion.ltsFrom.isSameOrBefore(date) && nodeVersion.to.isSameOrAfter(date)
      )
      .sort(dateSorter(`ltsFrom`))
      .map((nodeVersion: INodeVersion) => nodeVersion.version);
    const latest = relevantVersions[relevantVersions.length - 1];
    if (!latest) {
      this.logger.error(`Failed to resolve latest lts version in ${date.toJSON()}`);
      return;
    }
    this.logger.debug(`Resolved latest lts version in ${date.toJSON()} to ${latest}`);
    return latest;
  }

  @memoize(({ date }: INodeVersionsLatestOptions): string => date.toJSON())
  public async resolveStableVersion({ date }: INodeVersionsLatestOptions): Promise<string | undefined> {
    this.logger.debug(`Resolving stable version in ${date.toJSON()}`);
    const versions = await this.getAllVersions();
    const relevantVersions = versions
      .filter(
        (nodeVersion: INodeVersion) => nodeVersion.from.isSameOrBefore(date) && nodeVersion.to.isSameOrAfter(date)
      )
      .sort(dateSorter(`from`))
      .map((nodeVersion: INodeVersion) => nodeVersion.version);
    const latest = relevantVersions[relevantVersions.length - 1];
    if (!latest) {
      this.logger.error(`Failed to resolve stable version in ${date.toJSON()}`);
      return;
    }
    this.logger.debug(`Resolved stable version in ${date.toJSON()} to ${latest}`);
    return latest;
  }

  @memoize()
  private async getAllVersions(): Promise<INodeVersion[]> {
    this.logger.debug(`Fetching all node release versions`);
    try {
      const response = await this.axios.get<RemoteNodeVersions>(NODE_RELEASE_FILE);
      this.logger.debug(`Parsing versions result object`);
      return Object.entries(response.data)
        .filter(([, data]) => data.lts)
        .map(([version, data]) => {
          return {
            from: moment(data.start, dateFormat),
            to: moment(data.end, dateFormat),
            ltsFrom: moment(data.lts, dateFormat),
            version: version.substr(1),
            codename: (data.codename || ``).toLowerCase(),
          };
        });
    } catch (err) {
      this.logger.error(`Failed to fetch node release versions`, err);
      throw new Error(`Failed to fetch node release versions`);
    }
  }
}
