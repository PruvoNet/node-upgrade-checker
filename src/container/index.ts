import { Container, ContainerModule, interfaces } from 'inversify';
import Bind = interfaces.Bind;
import { gitModuleBinder } from '../utils/git';
import { commandRunnerModuleBinder } from '../utils/commandRunner';
import { yarnModulesBinder } from '../utils/yarn';
import { ciResolverModulesBinder } from '../resolvers/ciResolver';
import { dbModulesBinder } from '../db';
import { cacheResolveModulesBinder } from '../resolvers/cacheResolver';
import { testResolverModulesBinder } from '../resolvers/testResolver';
import { dependencyCheckerModulesBinder } from '../dependencyChecker';
import { nodeVersionsModulesBinder } from '../utils/nodeVersions';
import { packageInfoModulesBinder } from '../utils/packageInfo';
import { nodeModulesBinder } from './nodeModulesContainer';
import { enginesResolveModulesBinder } from '../resolvers/enginesResolver';
import { loggerModuleBinder } from '../utils/logger';

export const container = new Container({
  skipBaseClassChecks: true,
});

type Binder = (bind: Bind) => void;

const binders: Binder[] = [
  gitModuleBinder,
  nodeModulesBinder,
  commandRunnerModuleBinder,
  yarnModulesBinder,
  ciResolverModulesBinder,
  dbModulesBinder,
  cacheResolveModulesBinder,
  testResolverModulesBinder,
  dependencyCheckerModulesBinder,
  nodeVersionsModulesBinder,
  packageInfoModulesBinder,
  enginesResolveModulesBinder,
  loggerModuleBinder,
];

container.load(
  new ContainerModule((bind) => {
    binders.forEach((binder) => {
      binder(bind);
    });
  })
);
