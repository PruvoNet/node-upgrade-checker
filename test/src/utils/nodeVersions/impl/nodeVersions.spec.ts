import moment = require('moment');
import { NodeVersions } from '../../../../../src/utils/nodeVersions/impl/nodeVersions';
import { Axios } from '../../../../../src/container/nodeModulesContainer';
import { mock, mockReset } from 'jest-mock-extended';
import { ILogger } from '../../../../../src/utils/logger/interfaces/ILogger';
import { ILoggerFactory } from '../../../../../src/utils/logger';

const dateFormat = `YYYY-MM-DD`;

describe(`node versions`, () => {
  const nodeReleaseVersions = {
    'v0.10': {
      start: `2013-03-11`,
      end: `2016-10-31`,
    },
    'v0.12': {
      start: `2015-02-06`,
      end: `2016-12-31`,
    },
    v4: {
      start: `2015-09-08`,
      lts: `2015-10-12`,
      maintenance: `2017-04-01`,
      end: `2018-04-30`,
      codename: `Argon`,
    },
    v5: {
      start: `2015-10-29`,
      maintenance: `2016-04-30`,
      end: `2016-06-30`,
    },
    v6: {
      start: `2016-04-26`,
      lts: `2016-10-18`,
      maintenance: `2018-04-30`,
      end: `2019-04-30`,
      codename: `Boron`,
    },
    v7: {
      start: `2016-10-25`,
      maintenance: `2017-04-30`,
      end: `2017-06-30`,
    },
    v8: {
      start: `2017-05-30`,
      lts: `2017-10-31`,
      maintenance: `2019-01-01`,
      end: `2019-12-31`,
      codename: `Carbon`,
    },
    v9: {
      start: `2017-10-01`,
      maintenance: `2018-04-01`,
      end: `2018-06-30`,
    },
    v10: {
      start: `2018-04-24`,
      lts: `2018-10-30`,
      maintenance: `2020-04-30`,
      end: `2021-04-30`,
      codename: `Dubnium`,
    },
    v11: {
      start: `2018-10-23`,
      maintenance: `2019-04-22`,
      end: `2019-06-01`,
    },
    v12: {
      start: `2019-04-23`,
      lts: `2019-10-21`,
      maintenance: `2020-10-20`,
      end: `2022-04-30`,
      codename: `Erbium`,
    },
    v13: {
      start: `2019-10-22`,
      maintenance: `2020-04-01`,
      end: `2020-06-01`,
    },
    v14: {
      start: `2020-04-21`,
      lts: `2020-10-20`,
      maintenance: `2021-10-19`,
      end: `2023-04-30`,
      codename: ``,
    },
  };

  const axiosMock = mock<Axios>();
  let nodeVersions: NodeVersions;
  const loggerMock = mock<ILogger>();
  const loggerFactoryMock = mock<ILoggerFactory>();
  loggerFactoryMock.getLogger.mockReturnValue(loggerMock);

  beforeEach(() => {
    mockReset(loggerMock);
    mockReset(axiosMock);
    axiosMock.get.mockResolvedValue({
      data: nodeReleaseVersions,
    });
    nodeVersions = new NodeVersions(axiosMock, loggerFactoryMock);
  });

  describe(`stable`, () => {
    it(`should throw if fails to locate versions from remote url`, async () => {
      mockReset(axiosMock);
      axiosMock.get.mockRejectedValue(new Error(`failed to fetch`));
      const promise = nodeVersions.resolveStableVersion({
        date: moment.utc(`2015-10-02`, dateFormat),
      });
      await expect(promise).rejects.toBeInstanceOf(Error);
      await expect(promise).rejects.toMatchObject({
        message: `Failed to fetch node release versions`,
      });
      expect(axiosMock.get).toBeCalledTimes(1);
      expect(axiosMock.get).toHaveBeenCalledWith(
        `https://raw.githubusercontent.com/nodejs/Release/master/schedule.json`
      );
    });

    it(`should cache remote versions result`, async () => {
      await nodeVersions.resolveStableVersion({
        date: moment.utc(`2015-10-02`, dateFormat),
      });
      await nodeVersions.resolveStableVersion({
        date: moment.utc(`2015-10-03`, dateFormat),
      });
      expect(axiosMock.get).toBeCalledTimes(1);
      expect(axiosMock.get).toHaveBeenCalledWith(
        `https://raw.githubusercontent.com/nodejs/Release/master/schedule.json`
      );
    });

    it(`should cache versions result`, async () => {
      const res1 = await nodeVersions.resolveStableVersion({
        date: moment.utc(`2015-10-02`, dateFormat),
      });
      const res2 = await nodeVersions.resolveStableVersion({
        date: moment.utc(`2015-10-02`, dateFormat),
      });
      const res3 = await nodeVersions.resolveStableVersion({
        date: moment.utc(`2017-10-03`, dateFormat),
      });
      expect(res1).toBe(res2);
      expect(res2).not.toBe(res3);
      expect(axiosMock.get).toBeCalledTimes(1);
      expect(axiosMock.get).toHaveBeenCalledWith(
        `https://raw.githubusercontent.com/nodejs/Release/master/schedule.json`
      );
    });

    it(`should resolve to node 4`, async () => {
      const versions = await nodeVersions.resolveStableVersion({
        date: moment.utc(`2015-10-02`, dateFormat),
      });
      expect(versions).toEqual(`4`);
    });

    it(`should resolve to node 6`, async () => {
      const versions = await nodeVersions.resolveStableVersion({
        date: moment.utc(`2016-10-19`, dateFormat),
      });
      expect(versions).toEqual(`6`);
    });

    it(`should resolve to node 8`, async () => {
      const versions = await nodeVersions.resolveStableVersion({
        date: moment.utc(`2017-11-01`, dateFormat),
      });
      expect(versions).toEqual(`8`);
    });

    it(`should resolve to node 10`, async () => {
      const versions = await nodeVersions.resolveStableVersion({
        date: moment.utc(`2018-11-01`, dateFormat),
      });
      expect(versions).toEqual(`10`);
    });

    it(`should resolve to node 12`, async () => {
      const versions = await nodeVersions.resolveStableVersion({
        date: moment.utc(`2019-10-23`, dateFormat),
      });
      expect(versions).toEqual(`12`);
    });

    it(`should resolve to node 14`, async () => {
      const versions = await nodeVersions.resolveStableVersion({
        date: moment.utc(`2020-10-21`, dateFormat),
      });
      expect(versions).toEqual(`14`);
    });

    it(`should not resolve`, async () => {
      const versions = await nodeVersions.resolveStableVersion({
        date: moment.utc(`2050-10-21`, dateFormat),
      });
      expect(versions).toEqual(undefined);
      expect(loggerMock.error).toHaveBeenCalledWith(`Failed to resolve stable version in 2050-10-21T00:00:00.000Z`);
    });
  });

  describe(`latest lts`, () => {
    it(`should throw if fails to locate versions from remote url`, async () => {
      mockReset(axiosMock);
      axiosMock.get.mockRejectedValue(new Error(`failed to fetch`));
      const promise = nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2015-10-02`, dateFormat),
      });
      await expect(promise).rejects.toBeInstanceOf(Error);
      await expect(promise).rejects.toMatchObject({
        message: `Failed to fetch node release versions`,
      });
      expect(axiosMock.get).toBeCalledTimes(1);
      expect(axiosMock.get).toHaveBeenCalledWith(
        `https://raw.githubusercontent.com/nodejs/Release/master/schedule.json`
      );
    });

    it(`should cache remote versions result`, async () => {
      const res1 = await nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2015-10-02`, dateFormat),
      });
      const res2 = await nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2015-10-02`, dateFormat),
      });
      const res3 = await nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2017-10-03`, dateFormat),
      });
      expect(res1).toBe(res2);
      expect(res2).not.toBe(res3);
      expect(axiosMock.get).toBeCalledTimes(1);
      expect(axiosMock.get).toHaveBeenCalledWith(
        `https://raw.githubusercontent.com/nodejs/Release/master/schedule.json`
      );
    });

    it(`should cache versions result`, async () => {
      await nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2015-10-02`, dateFormat),
      });
      await nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2015-10-02`, dateFormat),
      });
      expect(axiosMock.get).toBeCalledTimes(1);
      expect(axiosMock.get).toHaveBeenCalledWith(
        `https://raw.githubusercontent.com/nodejs/Release/master/schedule.json`
      );
    });

    it(`should resolve to node 4`, async () => {
      const versions = await nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2016-04-27`, dateFormat),
      });
      expect(versions).toEqual(`4`);
    });

    it(`should resolve to node 6`, async () => {
      const versions = await nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2017-06-01`, dateFormat),
      });
      expect(versions).toEqual(`6`);
    });

    it(`should resolve to node 8`, async () => {
      const versions = await nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2018-04-25`, dateFormat),
      });
      expect(versions).toEqual(`8`);
    });

    it(`should resolve to node 10`, async () => {
      const versions = await nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2019-04-24`, dateFormat),
      });
      expect(versions).toEqual(`10`);
    });

    it(`should resolve to node 12`, async () => {
      const versions = await nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2020-04-25`, dateFormat),
      });
      expect(versions).toEqual(`12`);
    });

    it(`should resolve to node 14`, async () => {
      const versions = await nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2020-10-21`, dateFormat),
      });
      expect(versions).toEqual(`14`);
    });

    it(`should not resolve`, async () => {
      const versions = await nodeVersions.resolveLatestLtsVersion({
        date: moment.utc(`2050-10-21`, dateFormat),
      });
      expect(versions).toEqual(undefined);
      expect(loggerMock.error).toHaveBeenCalledWith(`Failed to resolve latest lts version in 2050-10-21T00:00:00.000Z`);
    });
  });

  describe(`lts`, () => {
    it(`should throw if fails to locate versions from remote url`, async () => {
      mockReset(axiosMock.get);
      axiosMock.get.mockRejectedValue(new Error(`failed to fetch`));
      const promise = nodeVersions.resolveLtsVersion({
        codename: `Argon`,
      });
      await expect(promise).rejects.toBeInstanceOf(Error);
      await expect(promise).rejects.toMatchObject({
        message: `Failed to fetch node release versions`,
      });
      expect(axiosMock.get).toBeCalledTimes(1);
      expect(axiosMock.get).toHaveBeenCalledWith(
        `https://raw.githubusercontent.com/nodejs/Release/master/schedule.json`
      );
    });

    it(`should cache remote versions result`, async () => {
      await nodeVersions.resolveLtsVersion({
        codename: `Argon`,
      });
      await nodeVersions.resolveLtsVersion({
        codename: `argon`,
      });
      expect(axiosMock.get).toBeCalledTimes(1);
      expect(axiosMock.get).toHaveBeenCalledWith(
        `https://raw.githubusercontent.com/nodejs/Release/master/schedule.json`
      );
    });

    it(`should cache versions result`, async () => {
      const res1 = await nodeVersions.resolveLtsVersion({
        codename: `Argon`,
      });
      const res2 = await nodeVersions.resolveLtsVersion({
        codename: `Argon`,
      });
      const res3 = await nodeVersions.resolveLtsVersion({
        codename: `Boron`,
      });
      expect(res1).toBe(res2);
      expect(res2).not.toBe(res3);
      expect(axiosMock.get).toBeCalledTimes(1);
      expect(axiosMock.get).toHaveBeenCalledWith(
        `https://raw.githubusercontent.com/nodejs/Release/master/schedule.json`
      );
    });

    it(`should resolve to node 4 ignore case`, async () => {
      const versions = await nodeVersions.resolveLtsVersion({
        codename: `Argon`,
      });
      expect(versions).toEqual(`4`);
    });

    it(`should resolve to node 4`, async () => {
      const versions = await nodeVersions.resolveLtsVersion({
        codename: `argon`,
      });
      expect(versions).toEqual(`4`);
    });

    it(`should resolve to node 6`, async () => {
      const versions = await nodeVersions.resolveLtsVersion({
        codename: `Boron`,
      });
      expect(versions).toEqual(`6`);
    });

    it(`should resolve to node 8`, async () => {
      const versions = await nodeVersions.resolveLtsVersion({
        codename: `Carbon`,
      });
      expect(versions).toEqual(`8`);
    });

    it(`should resolve to node 10`, async () => {
      const versions = await nodeVersions.resolveLtsVersion({
        codename: `Dubnium`,
      });
      expect(versions).toEqual(`10`);
    });

    it(`should resolve to node 12`, async () => {
      const versions = await nodeVersions.resolveLtsVersion({
        codename: `Erbium`,
      });
      expect(versions).toEqual(`12`);
    });

    it(`should not resolve`, async () => {
      const versions = await nodeVersions.resolveLtsVersion({
        codename: `dummy`,
      });
      expect(versions).toEqual(undefined);
      expect(loggerMock.error).toHaveBeenCalledWith(`Failed to resolve LTS version of dummy`);
    });
  });
});
