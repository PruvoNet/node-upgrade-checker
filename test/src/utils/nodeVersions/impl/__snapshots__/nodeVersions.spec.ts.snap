// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`node versions getAllVersions should parse remote versions 1`] = `
Array [
  Object {
    "codename": "argon",
    "from": "2015-09-08T00:00:00.000Z",
    "ltsFrom": "2015-10-12T00:00:00.000Z",
    "to": "2018-04-30T00:00:00.000Z",
    "version": "4",
  },
  Object {
    "codename": "boron",
    "from": "2016-04-26T00:00:00.000Z",
    "ltsFrom": "2016-10-18T00:00:00.000Z",
    "to": "2019-04-30T00:00:00.000Z",
    "version": "6",
  },
  Object {
    "codename": "carbon",
    "from": "2017-05-30T00:00:00.000Z",
    "ltsFrom": "2017-10-31T00:00:00.000Z",
    "to": "2019-12-31T00:00:00.000Z",
    "version": "8",
  },
  Object {
    "codename": "dubnium",
    "from": "2018-04-24T00:00:00.000Z",
    "ltsFrom": "2018-10-30T00:00:00.000Z",
    "to": "2021-04-30T00:00:00.000Z",
    "version": "10",
  },
  Object {
    "codename": "erbium",
    "from": "2019-04-23T00:00:00.000Z",
    "ltsFrom": "2019-10-21T00:00:00.000Z",
    "to": "2022-04-30T00:00:00.000Z",
    "version": "12",
  },
  Object {
    "codename": "",
    "from": "2020-04-21T00:00:00.000Z",
    "ltsFrom": "2020-10-20T00:00:00.000Z",
    "to": "2023-04-30T00:00:00.000Z",
    "version": "14",
  },
]
`;
